noir-projects/noir-contracts/contracts/key_registry_contract/src/main.nr

```diff
@@ -0,0 +1,144 @@
+contract KeyRegistry {
+  use dep::aztec::{
+    avm::hash::poseidon,
+    context::gas::GasOpts,
+    state_vars::{
+      SharedMutable,
+      Map
+    }, 
+    protocol_types::{
+      abis::function_selector::FunctionSelector,
+      contract_class_id::ContractClassId,
+      address::{
+        AztecAddress,
+        EthAddress,
+        PublicKeysHash,
+        PartialAddress,
+      },
+      constants::{
+        GENERATOR_INDEX__CONTRACT_ADDRESS,
+        GENERATOR_INDEX__PARTIAL_ADDRESS
+      },
+    },
+  };
+
+  global KEY_ROTATION_DELAY = 5;
+
+  #[aztec(storage)]
+  struct Storage {
+    nullifier_public_key_registry: Map<AztecAddress, SharedMutable<Field, KEY_ROTATION_DELAY>>,
+    incoming_public_key_registry: Map<AztecAddress, SharedMutable<Field, KEY_ROTATION_DELAY>>,
+    outgoing_public_key_registry: Map<AztecAddress, SharedMutable<Field, KEY_ROTATION_DELAY>>,
+    tagging_public_key_registry: Map<AztecAddress, SharedMutable<Field, KEY_ROTATION_DELAY>>,
+  }
+
+  #[aztec(public)]
+  fn register(
+    nullifier_public_key: Field,
+    incoming_public_key: Field,
+    outgoing_public_key: Field,
+    tagging_public_key: Field,
+  ) {
+    assert((nullifier_public_key != 0) & (incoming_public_key != 0) & (outgoing_public_key != 0) & (tagging_public_key != 0));
+    let nullifier_entry = storage.nullifier_public_key_registry.at(context.msg_sender());
+    let incoming_entry = storage.nullifier_public_key_registry.at(context.msg_sender());
+    let outgoing_entry = storage.nullifier_public_key_registry.at(context.msg_sender());
+    let tagging_entry = storage.nullifier_public_key_registry.at(context.msg_sender());
+
+    nullifier_entry.schedule_value_change(nullifier_public_key);
+    incoming_entry.schedule_value_change(incoming_public_key);
+    outgoing_entry.schedule_value_change(outgoing_public_key);
+    tagging_entry.schedule_value_change(tagging_public_key);
+  }
+
+  #[aztec(public)]
+  fn register_from_preimage(
+    address: AztecAddress,
+    contract_class_id: ContractClassId,
+    initialization_hash: Field,
+    salt: Field,
+    deployer: AztecAddress,
+    portal_contract_address: EthAddress,
```

Comment:
the 5 lines above should be replaced just with partial address. The goal of this check is to ensure that nobody can register invalid keys for an address and just having a partial address here achieves that.